examples/cmd/pool.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/src-d/metadata-retrieval/github"
+)
+
+type DownloadersPool struct {
+	Size    int
+	pool    chan *DownloaderPoolItem
+	started bool
+	ended   bool
+	t0      time.Time
+	t1      time.Time
+	stats0  map[int]*downloaderStats
+	stats1  map[int]*downloaderStats
+}
+
+type DownloaderPoolItem struct {
+	id int
+	*github.Downloader
+}
+
+type DownloaderPoolStats struct {
+	Elapsed    time.Duration
+	RatesUsage map[int]*RateUsage
+}
+
+func (d *DownloaderPoolItem) stats() (*downloaderStats, error) {
+	rate, err := d.RateRemaining(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	return &downloaderStats{
+		Rate: rate,
+		Time: time.Now(),
+	}, nil
+}
+
+type downloaderStats struct {
+	Rate int
+	Time time.Time
+}
+
+type downloaderBuilder = func(c *http.Client) (*github.Downloader, error)
+
+func NewDownloadersPool(clients []*http.Client, builder downloaderBuilder) (*DownloadersPool, error) {
+	ch := make(chan *DownloaderPoolItem, len(clients))
+
+	for i, c := range clients {
+		d, err := builder(c)
+		if err != nil {
+			return nil, err
+		}
+
+		ch <- &DownloaderPoolItem{id: i, Downloader: d}
+	}
+
+	return &DownloadersPool{
+		Size: len(clients),
+		pool: ch,
+	}, nil
+}
+
+func (dp *DownloadersPool) WithDownloader(f func(d *DownloaderPoolItem) error) error {
+	if !dp.started || dp.ended {
```

Comment:
could you please explain me what is the advantage of adding `started` field and `Begin` method instead of starting it in the constructor?

Reply:
The idea is to separate initialization from from when timing starts.

Reply:
true that in this case we actually call `Begin` immediately after constructor. I guess that I'm thinking by habit about flexibility when in this case it's going to be actually used by this cli.

Reply:
yeah. I don't mind keeping it as is. It was just surprising to see it in this particular case.